synnax/pkg/distribution/channel/verification/service.go

```diff
@@ -0,0 +1,215 @@
+
+	var ctx context.Context
+	if key == "" {
+		return service, nil
+	}
+
+	// user did enter a key
+	err = service.create(ctx, key)
+	if err != nil {
+		return service, err
+	}
+	sCtx.Go(service.updateKeyValidation)
+
+	return service, err
+}
+
+// Close will shutdown the service
+func (s *Service) Close() error {
+	return s.shutdown.Close()
+}
+
+// ValidateChannelCount returns an error if inUse is greater than the number of
+// channels the user is allowed to use.
+func (s *Service) ValidateChannelCount(ctx context.Context, inUse int64) error {
+	key, err := s.retrieve(ctx)
+
+	if err != nil {
+		if inUse > maxFreeChannels {
+			return errFreeLimit(maxFreeChannels)
+		}
+		return nil
+	}
+
+	if getExpirationDate(key).Before(time.Now()) {
+		if inUse > maxFreeChannels {
+			return errExpireKey(maxFreeChannels)
+		}
+		return nil
+	}
+
+	if channelsAllowed := getNumberOfChannels(key); inUse > channelsAllowed {
+		return errOverChannelLimit(int(channelsAllowed))
+	}
+
+	return nil
+}
+
+// IsExpired returns an error if the key stored in s is expired.
+func (s *Service) IsExpired(ctx context.Context) error {
+	key, err := s.retrieve(ctx)
+	if err != nil {
+		return errFreeLimit(maxFreeChannels)
+
+	}
+	if getExpirationDate(key).Before(time.Now()) {
+		return errExpireKey(maxFreeChannels)
+	}
+	return nil
+}
+
+func (s *Service) GetMaxFreeChannels() int {
+	return maxFreeChannels
+}
+
+// create sets  key in s.DB. Returns an error if the key is invalid - you can
+// only set a valid key in the database
+func (s *Service) create(ctx context.Context, key string) error {
+	err := validateKey(key)
+	if err != nil {
+		return err
+	}
+	return s.DB.Set(ctx, []byte("productKey"), []byte(key))
+}
+
+// retrieve grabs the key in s.DB.
+func (s *Service) retrieve(ctx context.Context) (string, error) {
+	key, err := s.DB.Get(ctx, []byte("productKey"))
+	return string(key), err
+}
+
+// updateKeyValidation will run a background process that continuously checks if
+// the key is expired.
+func (s *Service) updateKeyValidation(ctx context.Context) error {
+
+	key, err := s.retrieve(ctx)
+	if err != nil {
+		return err
+	}
+	expireTime := getExpirationDate(key)
+	ticker := time.NewTicker(24 * time.Hour)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+			if expireTime.Before(time.Now()) {
+				s.Ins.L.Error("License key expired. Access has been limited", zap.String("expiredAt", expireTime.String()))
+			} else if timeLeft := time.Until(expireTime); timeLeft <= s.ExpireWarning {
+				s.Ins.L.Warn("License key will expire soon. Access will be limited.", zap.String("expiresIn", timeLeft.String()))
+			} else {
+				s.Ins.L.Info("License key is not expired.", zap.String("expiresIn", timeLeft.String()))
+			}
+		}
+	}
+}
+
+func errExpireKey(channelsAllowed int) error {
+	encoded1 := "dXNpbmcgYW4gZXhwaXJlZCBwcm9kdWN0IGxpY2Vuc2Uga2V5LCB1c2UgaXMgbGltaXRlZCB0byB0aGUgZmlyc3Qg"

```

Comment:
These encoded keys don't need to be defined within the function. they can be constants. In fact, the entire error can be extracted outside of this function.